agent/dns.go

```diff
@@ -270,7 +270,7 @@ func recursorAddr(recursor string) (string, error) {
 	// Add the port if none
 START:
 	_, _, err := net.SplitHostPort(recursor)
-	if ae, ok := err.(*net.AddrError); ok && ae.Err == "missing port in address" {
+	if ae, ok := err.(*net.AddrError); ok && (ae.Err == "missing port in address" || ae.Err == "too many colons in address") {
```

Comment:
I took a quick look at the implementation of `SplitHostPort` and there are several other real error cases where the `too many colons in address` can be returned. In those cases we would not want to use the `FormatAddressPort` logic with the default port but rather return the error.

I think something like the following should be sufficient to only assume a non-bracketed IPv6 address when we get that error.

```
if ae, ok := err.(*net.AddrError); ok {
  if ae.Err == "missing port in address" {
    recursor = ipaddr.FormatAddressPort(recursor, 53)
    goto START
  } else if ae.Err == "too many colons in address" {
    if ip := net.ParseIp(recursor); ip != nil && ip.To4() == nil {
      recursor = ipaddr.FormatAddressPort(recursor, 53)
      goto START
    }
  }
}
```

I haven't tested that myself but basically it just ensures that in the case of the too many colons error that the address is a valid IPv6 address and we didn't get that error for any other reason.

In addition, it would be great to also add a test for `foo::53` as that should return with the `too many colons in address` error but we should not attempt to turn it into an actual recursor.